numpy/core/src/umath/loops.c.src

```diff
@@ -913,9 +913,32 @@ NPY_NO_EXPORT void
 @TYPE@_power(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
 {
     BINARY_LOOP {
-        const @ftype@ in1 = (@ftype@)*(@type@ *)ip1;
-        const @ftype@ in2 = (@ftype@)*(@type@ *)ip2;
-        *((@type@ *)op1) = (@type@) pow(in1, in2);
+        @type@ in1 = *(@type@ *)ip1;
+        @type@ in2 = *(@type@ *)ip2;
+        @type@ out = 1;
+        @type@ temp;
+
+        if (in2 < 0 || in1 == 0) {
+            *((@type@ *)op1) = 0;
+            continue;
+        }
+        if (in2 == 0) {
+            *((@type@ *)op1) = 1;
+            continue;
+        }
+
+        while (in2 > 0) {
+            if (in2 & 1) {
+                out *= in1;
+            }
+            in2 >>= 1;        /* Shift exponent down by 1 bit */
+            if (in2 == 0) {
```

Comment:
Does this really speed things up at all? You are doing one check per iteration in order to save yourself a single multiplication in the last one, does it pay off.

If there is any performance to be gained there, you could alternatively peel off the first multiplication as an assignment, so that a single check is needed:

```C
out = in2 & 1 ? in1 : 1;
in2 >>= 1;
while (in2 > 0) {
    in1 *= in1;
    if (in2 & 1) {
        out *= in1;
    }
    in2 >>= 1;
}
```